internal/server/data/data.go

```diff
@@ -152,28 +139,69 @@ func add[T models.Modelable](db *gorm.DB, model *T) error {
 		return err
 	}
 
-	if err := db.Create(model).Error; err != nil {
-		if isUniqueConstraintViolation(err) {
-			return fmt.Errorf("%w: %s", internal.ErrDuplicate, err)
-		}
+	err := db.Create(model).Error
+	return handleError(err)
+}
 
-		return err
-	}
+type UniqueConstraintError struct {
+	Table  string
+	Column string
+}
 
-	return nil
+func (e UniqueConstraintError) Error() string {
+	if e.Table == "" {
+		return "value already exists"
+	} else if e.Column == "" {
+		return fmt.Sprintf("value already exists in %v", e.Table)
+	}
+	return fmt.Sprintf("value for %v already exists in %v", e.Column, e.Table)
```

Comment:
Very minor, not a blocker, but thinking about how the user will see this.

I wonder if "already exists for x" (ex: already exists for grant) would read better than "in".

Maybe not a concern if we're going to change the error later at the API or CLI layers.